File: packages/kolibri-build/src/collect_sandbox_static.js
const fs = require('node:fs');
const path = require('node:path');
const { gunzipSync } = require('node:zlib');
const logger = require('kolibri-logging');

const logging = logger.getLogger('Kolibri Sandbox Static Collector');

const GZ = '.gz';
const FILE_SIZE = '.file_size';

/**
 * @typedef {object} BundleDatum The fields of a read_webpack_json entry used here.
 * @property {string} module_path Dotted Python path of the plugin the bundle belongs to.
 * @property {string} plugin_path Filesystem path of the plugin package.
 * @property {string} static_dir The plugin's static directory.
 * @property {boolean} sandbox_handler Whether the bundle is a sandbox handler.
 */

/**
 * List the directories alt_wsgi.py mounts on the sandbox server: the core content
 * static directory (where the kolibri-sandbox package builds to) followed by the
 * static directory of every plugin shipping a sandbox handler bundle.
 * @param {BundleDatum[]} bundleData - All bundles read from the plugins being built.
 * @returns {string[]} Source directories, in the order the sandbox server resolves them.
 * @throws {Error} If kolibri.core is absent, leaving the core static path unknown.
 */
function sandboxStaticPaths(bundleData) {
  const core = bundleData.find(bundle => bundle.module_path === 'kolibri.core');
  if (!core) {
    throw new Error('kolibri.core must be in the plugin list to locate the core static files');
  }
  return [
    path.join(core.plugin_path, 'content', 'static'),
    ...new Set(bundleData.filter(bundle => bundle.sandbox_handler).map(b => b.static_dir)),
  ];
}

/**
 * Whether a file exists only to support serve-time reconstruction, so a plain static
 * host has no URL for it. DynamicWhiteNoise serves a .gz as content-encoding on its
 * primary's URL, never as a path of its own; a .gz with no primary is a content file.
 * @param {string} srcFile - Path of the candidate file.
 * @returns {boolean} True when the file should not be collected.
 */
function isServeArtifact(srcFile) {
  if (srcFile.endsWith(FILE_SIZE)) {
    return true;
  }
  return srcFile.endsWith(GZ) && fs.existsSync(srcFile.slice(0, -GZ.length));
}

/**
 * Copy one file, writing it whole. This undoes compress.js, which gzips each static
 * file, truncates the original to 0 bytes and leaves a .file_size sibling;
 * DynamicWhiteNoise rebuilds those at serve time but a CDN won't.
 * @param {string} srcFile - Path of the file to collect.
 * @param {string} destFile - Path to write, whose parent directory must already exist.
 */
function collectFile(srcFile, destFile) {
  const buildCompressed =
    fs.existsSync(srcFile + FILE_SIZE) &&
    fs.statSync(srcFile).size === 0 &&
    fs.existsSync(srcFile + GZ);
  if (buildCompressed) {
    fs.writeFileSync(destFile, gunzipSync(fs.readFileSync(srcFile + GZ)));
  } else {
    fs.copyFileSync(srcFile, destFile);
  }
}

/**
 * Resolve a path through symlinks, tolerating one that does not exist yet by resolving
 * its nearest existing ancestor and re-appending the rest.
 * @param {string} target - Path to resolve, which need not exist.
 * @returns {string} Absolute path with every existing symlink component resolved.
 */
function realPath(target) {
  const missing = [];
  let existing = path.resolve(target);
  while (!fs.existsSync(existing)) {
    missing.unshift(path.basename(existing));
    existing = path.dirname(existing);
  }
  return path.join(fs.realpathSync(existing), ...missing);
}

/**
 * Refuse a destination that overlaps a source tree, in either direction: one containing
 * a source loses it to --clear, and one nested inside a source is walked as it is
 * written, collecting each file again under a doubled relative path.
 * @param {string} destination - Directory to collect into, which need not exist yet.
 * @param {string[]} sourceDirs - Source directories to collect from.
 * @throws {Error} If the destination overlaps a source directory.
 */
function assertDisjoint(destination, sourceDirs) {
  const destReal = realPath(destination);
  for (const sourceDir of sourceDirs) {
    if (!fs.existsSync(sourceDir)) {
      continue;
    }
    const sourceReal = fs.realpathSync(sourceDir);
    if (sourceReal === destReal || sourceReal.startsWith(destReal + path.sep)) {
      throw new Error(`Refusing to collect into ${destination}: it contains source ${sourceDir}`);
    }
    if (destReal.startsWith(sourceReal + path.sep)) {
      throw new Error(`Refusing to collect into ${destination}: it is inside source ${sourceDir}`);
    }
  }
}

/**
 * Delete an existing destination tree, refusing a symlink: rm would remove the link
 * rather than the tree it names, leaving the stale files in place.
 * @param {string} destination - Directory to delete, which must exist.
 * @throws {Error} If the destination is a symlink.
 */
function clearDestination(destination) {
  if (fs.lstatSync(destination).isSymbolicLink()) {
    throw new Error(`Refusing to clear symlinked destination: ${destination}`);
  }
  logging.info(`Clearing destination directory: ${destination}`);
  fs.rmSync(destination, { recursive: true });
}

/**
 * Recursively yield every file under a directory.
 * @param {string} dir - Directory to walk.
 * @yields {string} Path of each file found, at any depth.
 */
function* walkFiles(dir) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const entryPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      yield* walkFiles(entryPath);
    } else {
      yield entryPath;
    }
  }
}

/**
 * Copy every file the sandbox server serves into a single directory, suitable for
 * upload to a CDN or object storage.
 * @param {BundleDatum[]} bundleData - All bundles read from the plugins being built.
 * @param {string} destination - Directory to collect into; created if absent.
 * @param {object} options - Collection options.
 * @param {boolean} [options.clear] - Delete the destination tree before collecting.
 * @returns {number} How many files were collected.
 * @throws {Error} If the sources yield no files, or the destination overlaps a source.
 */
function collectSandboxStatic(bundleData, destination, { clear = false } = {}) {
  const sourceDirs = sandboxStaticPaths(bundleData);

  assertDisjoint(destination, sourceDirs);

  if (clear && fs.existsSync(destination)) {
    clearDestination(destination);
  }

  fs.mkdirSync(destination, { recursive: true });

  const collected = new Set();

  for (const sourceDir of sourceDirs) {
    if (!fs.existsSync(sourceDir)) {
      logging.warn(`Source directory not found: ${sourceDir}`);
      continue;
    }

    logging.info(`Collecting from: ${sourceDir}`);

    for (const srcFile of walkFiles(sourceDir)) {
      if (isServeArtifact(srcFile)) {
        continue;
      }

      // FileFinder resolves colliding relative paths first-wins, in source order —
      // mirror that rather than overwriting.
      const relFile = path.relative(sourceDir, srcFile);
      if (collected.has(relFile)) {
        logging.debug(`Skipping ${srcFile}: already collected from an earlier source`);
        continue;
      }
      collected.add(relFile);

      const destFile = path.join(destination, relFile);
      fs.mkdirSync(path.dirname(destFile), { recursive: true });
      collectFile(srcFile, destFile);
    }
  }

  if (!collected.size) {
    throw new Error('No static files found to collect');
  }

  logging.info(`Successfully collected ${collected.size} files to ${destination}`);
  return collected.size;
}

module.exports = collectSandboxStatic;
module.exports.sandboxStaticPaths = sandboxStaticPaths;

File: build_tools/collect_sandbox_static.py
"""
Collect the static files the sandbox server serves into one directory, for upload to
object storage. Run from the project environment, after the frontend is built:

    uv run python build_tools/collect_sandbox_static.py <destination>
"""

import argparse
import gzip
import os
import shutil
import sys
import tempfile

GZ = ".gz"
FILE_SIZE = ".file_size"


def sandbox_sources():
    """
    The directories alt_wsgi.py mounts on the sandbox server, in resolution order, and
    the built output directory of every sandbox handler bundle.
    """
    # kolibri.utils.conf reads KOLIBRI_HOME at import, so the caller sets it first.
    from kolibri.core.content.hooks import SandboxedContentViewerHook
    from kolibri.plugins.registry import registered_plugins

    list(registered_plugins)
    handler_outputs = [
        os.path.join(hook.sandbox_static_path, hook.sandbox_handler_unique_id)
        for hook in SandboxedContentViewerHook.registered_hooks
    ]
    return SandboxedContentViewerHook.get_sandbox_static_paths(), handler_outputs


def _is_serve_artifact(path):
    # DynamicWhiteNoise serves a .gz as content-encoding on its primary's URL, so a
    # static host has no URL for it; a .gz with no primary is a content file.
    if path.endswith(FILE_SIZE):
        return True
    return path.endswith(GZ) and os.path.exists(path[: -len(GZ)])


def _collect_file(src, dest):
    # compress.js truncates each file to 0 bytes beside its .gz and .file_size;
    # DynamicWhiteNoise rebuilds it at serve time, a CDN won't.
    compressed = (
        os.path.exists(src + FILE_SIZE)
        and os.path.getsize(src) == 0
        and os.path.exists(src + GZ)
    )
    if compressed:
        with gzip.open(src + GZ, "rb") as f_in, open(dest, "wb") as f_out:
            shutil.copyfileobj(f_in, f_out)
    else:
        shutil.copyfile(src, dest)


def _assert_built(source_dirs, handler_outputs):
    for source_dir in source_dirs:
        if not os.path.isdir(source_dir):
            raise RuntimeError(
                f"Static directory not found, build it first: {source_dir}"
            )
    for output in handler_outputs:
        if not os.path.isdir(output) or not os.listdir(output):
            raise RuntimeError(
                f"Sandbox handler bundle not built, build it first: {output}"
            )


def _content_files(source_dir):
    for root, _, filenames in os.walk(source_dir):
        for filename in filenames:
            src = os.path.join(root, filename)
            if not _is_serve_artifact(src):
                yield src


def collect(source_dirs, handler_outputs, destination):
    _assert_built(source_dirs, handler_outputs)

    collected = {}
    for source_dir in source_dirs:
        for src in _content_files(source_dir):
            rel = os.path.relpath(src, source_dir)
            if rel in collected:
                raise RuntimeError(
                    f"{rel} is in both {collected[rel]} and {source_dir}; "
                    "a sandbox handler plugin must namespace its static files"
                )
            collected[rel] = source_dir
            dest = os.path.join(destination, rel)
            os.makedirs(os.path.dirname(dest), exist_ok=True)
            _collect_file(src, dest)

    if not collected:
        raise RuntimeError("No static files found to collect")
    return len(collected)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("destination")
    args = parser.parse_args()
    with tempfile.TemporaryDirectory() as home:
        # A fresh home enables exactly the default plugins.
        os.environ["KOLIBRI_HOME"] = home
        source_dirs, handler_outputs = sandbox_sources()
        count = collect(source_dirs, handler_outputs, args.destination)
    sys.stdout.write(f"Collected {count} files to {args.destination}\n")


if __name__ == "__main__":
    main()
